InnerEye/ML/SSL/lightning_containers/ssl_container.py

```diff
@@ -0,0 +1,243 @@
+        self.encoder_output_dim = get_encoder_output_dim(model, self.data_module)
+
+        return model
+
+    def get_data_module(self) -> InnerEyeDataModuleTypes:
+        """
+        Gets the data that is used for the training and validation steps.
+        Here we use different data loader for training of linear head and training of SSL model.
+        """
+        if hasattr(self, "data_module"):
+            return self.data_module
+        encoder_module = self._create_ssl_data_modules(is_ssl_encoder_module=True)
+        linear_head_module = self._create_ssl_data_modules(is_ssl_encoder_module=False)
+        return CombinedDataModule(encoder_module, linear_head_module, self.use_balanced_binary_loss_for_linear_head)
+
+    def _create_ssl_data_modules(self, is_ssl_encoder_module: bool) -> InnerEyeVisionDataModule:
+        """
+        Returns torch lightning data module for encoder or linear head
+
+        :param is_ssl_encoder_module: whether to return the data module for SSL training or for linear heard. If true,
+        :return transforms with two views per sample (batch like (img_v1, img_v2, label)). If False, return only one
+        view per sample but also return the index of the sample in the dataset (to make sure we don't use twice the same
+        batch in one training epoch (batch like (index, img_v1, label), as classifier dataloader expected to be shorter
+        than SSL training, hence CombinedDataloader might loop over data several times per epoch).
+        """
+        datamodule_args = self.datamodule_args[SSLDataModuleType.ENCODER] if is_ssl_encoder_module else \
+            self.datamodule_args[SSLDataModuleType.LINEAR_HEAD]
+
+        train_transforms, val_transforms = self._get_transforms(datamodule_args.augmentation_params,
+                                                                datamodule_args.dataset_name,
+                                                                is_ssl_encoder_module)
+        batch_size_per_gpu = datamodule_args.batch_size // self.total_num_gpus if self.total_num_gpus > 0 else \
+            datamodule_args.batch_size
+        logging.info(f"Batch size per gpu: {batch_size_per_gpu}")
+        dm = InnerEyeVisionDataModule(dataset_cls=self._SSLDataClassMappings[datamodule_args.dataset_name],
+                                      return_index=not is_ssl_encoder_module,  # index is only needed for linear head
+                                      train_transforms=train_transforms,
+                                      val_split=0.1,
+                                      val_transforms=val_transforms,
+                                      data_dir=str(datamodule_args.dataset_path),
+                                      batch_size=batch_size_per_gpu,
+                                      num_workers=self.num_workers,
+                                      seed=self.random_seed)
+        dm.prepare_data()
+        dm.setup('fit')
```

Comment:
Please ignore this comment if we are using this function only at training time. Otherwise, maybe it's better to think if it actually changes the behaviour at test time.

Reply:
Test time is not implemented for ssl_training. But in anycase we should just call setup on it, the parameter 'fit' is not needed you are right